sanctum-solana-client-utils/src/prioritization_fee.rs

```diff
@@ -0,0 +1,154 @@
+use std::cmp::min;
+
+use medians::Medianf64;
+use solana_client::{
+    nonblocking::rpc_client::RpcClient as NonblockingRpcClient, rpc_client::RpcClient,
+    rpc_response::RpcPrioritizationFee,
+};
+use solana_rpc_client_api::{
+    client_error::Error as ClientError, config::RpcSimulateTransactionConfig,
+};
+use solana_sdk::{
+    compute_budget::ComputeBudgetInstruction, instruction::Instruction, message::Message,
+    pubkey::Pubkey, transaction::Transaction,
+};
+
+const MAX_SLOT_DISPLACEMENT: u64 = 150;
+const WEIGHTED_MEDIAN_EPSILON: f64 = 0.0001;
+
+pub fn get_compute_budget_ixs(unit_limit: u32, unit_price_micro_lamports: u64) -> [Instruction; 2] {
+    [
+        ComputeBudgetInstruction::set_compute_unit_limit(unit_limit),
+        ComputeBudgetInstruction::set_compute_unit_price(unit_price_micro_lamports),
+    ]
+}
+
+// assumes <= `MAX_SLOT_DISPLACEMENT` slots of sample size
+pub fn calc_slot_weighted_median_prioritization_fees(
+    rpc_prio_fees: &[RpcPrioritizationFee],
+) -> u64 {
+    if rpc_prio_fees.is_empty() {
+        log::warn!("Given sample for the recent prioritization fee was empty");
+        return 0u64;
+    }
+
+    // min_slot (the most recent slot) is assumed to be max - MAX_SLOT_DISPLACEMENT
+    // NOTE: data for the slot at max - MAX_SLOT_DISPLACEMENT might be missing but we
+    // assume that it exists and the fee was 0 for that slot
+    let min_slot = rpc_prio_fees
+        .iter()
+        .max_by_key(|fee| fee.slot)
+        .unwrap()
+        .slot
+        - MAX_SLOT_DISPLACEMENT;
+
+    let (values, weights): (Vec<f64>, Vec<f64>) = rpc_prio_fees
+        .iter()
+        .filter_map(|fee| {
+            if fee.prioritization_fee == 0 {
+                None
+            } else {
+                Some((
+                    fee.prioritization_fee as f64,
+                    (1 + fee.slot - min_slot) as f64 / MAX_SLOT_DISPLACEMENT as f64,
```

Comment:
might you get 0 or negative weights here? I know currently the docs say its up to 150 blocks but it doesnt say if its 150 contiguous blocks or how it deals with skipped slots does it? Guess we can filter older stuff out